File: min-csv-rnn/csv-rnn.py
"""
Minimal character-level Vanilla RNN model. Written by Andrej Karpathy (@karpathy)
BSD License
"""
import numpy as np
import csv
from tabulate import tabulate

# data I/O
file = csv.reader(open('cyclic.csv','r'))
features = []
for row in file:
  featuresRow = []
  for element in row:
    featuresRow.append(float(element))
  features.append(featuresRow)

amountOfFrames = len(features)
amountOfSamples = len(features[0])
  
print 'data has %d frames, %d samples.' % (amountOfFrames, amountOfSamples)

# hyperparameters
hidden_size = 100 # size of hidden layer of neurons
seq_length = 25 # number of steps to unroll the RNN for
learning_rate = 1e-1

# model parameters
Wxh = np.random.randn(hidden_size, amountOfSamples)*0.01 # input to hidden
Whh = np.random.randn(hidden_size, hidden_size)*0.01 # hidden to hidden
Why = np.random.randn(amountOfSamples, hidden_size)*0.01 # hidden to output
bh = np.zeros((hidden_size, 1)) # hidden bias
by = np.zeros((amountOfSamples, 1)) # output bias

def lossFun(inputs, targets, hprev):
  """
  inputs,targets are arrays that contain 25 frames.
  hprev is Hx1 array of initial hidden state
  returns the loss, gradients on model parameters, and last hidden state
  """
  xs, hs, ys, ps, diff = {}, {}, {}, {}, {}
  hs[-1] = np.copy(hprev)
  loss = 0
  # forward pass
  for t in xrange(len(inputs)):
    xs[t] = np.array([inputs[t]]).T
    hs[t] = np.tanh(np.dot(Wxh, xs[t]) + np.dot(Whh, hs[t-1]) + bh) # hidden layer activations
    ys[t] = np.dot(Why, hs[t]) + by # predictions
    #ys[t][ys[t] < 0] = 0 # predictions must be > 0
    diff[t] = np.subtract(np.array([targets[t]]).T, ys[t])
    loss += np.dot(diff[t].T,diff[t])[0][0]

  # backward pass: compute gradients going backwards
  dWxh, dWhh, dWhy = np.zeros_like(Wxh), np.zeros_like(Whh), np.zeros_like(Why)
  dbh, dby = np.zeros_like(bh), np.zeros_like(by)
  dhnext = np.zeros_like(hs[0])
  for t in reversed(xrange(len(inputs))):
    dy = ys[t] - np.array([targets[t]]).T
    dWhy += np.dot(dy, hs[t].T)
    dby += dy
    dh = np.dot(Why.T, dy) + dhnext # backprop into h
    dhraw = (1 - hs[t] * hs[t]) * dh # backprop through tanh nonlinearity
    dbh += dhraw
    dWxh += np.dot(dhraw, xs[t].T)
    dWhh += np.dot(dhraw, hs[t-1].T)
    dhnext = np.dot(Whh.T, dhraw)
  for dparam in [dWxh, dWhh, dWhy, dbh, dby]:
    np.clip(dparam, -5, 5, out=dparam) # clip to mitigate exploding gradients
  return loss, dWxh, dWhh, dWhy, dbh, dby, hs[len(inputs)-1]

def sample(h, seed, n):
  x = np.array([seed]).T
  samples = []
  for t in xrange(n):
    h = np.tanh(np.dot(Wxh, x) + np.dot(Whh, h) + bh)
    y = np.dot(Why, h) + by
    samples.append(y.T[0])
    x = y
  print tabulate(np.round(samples,4))

n, p = 0, 0
mWxh, mWhh, mWhy = np.zeros_like(Wxh), np.zeros_like(Whh), np.zeros_like(Why)
mbh, mby = np.zeros_like(bh), np.zeros_like(by) # memory variables for Adagrad
smooth_loss = -np.log(1.0/amountOfSamples)*seq_length # loss at iteration 0
hprev = np.zeros((hidden_size,1)) # reset RNN memory

def humanizeLargeArray(A):
  smallerA = []
  for i in xrange(0,10):
    smallerA.append(A[i][0:10])
  return smallerA

while True:
  # prepare inputs (we're sweeping from left to right in steps seq_length long)
  if p+seq_length+1 >= amountOfFrames or n == 0: 
    p = 0 # go from start of data
    hprev = np.zeros((hidden_size,1)) # reset RNN memory

  inputs = features[p:p+seq_length]
  targets = features[p+1:p+seq_length+1]

  # forward seq_length characters through the net and fetch gradient
  loss, dWxh, dWhh, dWhy, dbh, dby, hprev = lossFun(inputs, targets, hprev)
  smooth_loss = smooth_loss * 0.999 + loss * 0.001
  if n % 100 == 0: 
    print 'iter %d, loss: %f, pointer: %d' % (n, smooth_loss, p) # print progress
    sample(hprev,features[p],5)

  # perform parameter update with Adagrad
  for param, dparam, mem in zip([Wxh, Whh, Why, bh, by], 
                                [dWxh, dWhh, dWhy, dbh, dby], 
                                [mWxh, mWhh, mWhy, mbh, mby]):
    mem += dparam * dparam
    param += -learning_rate * dparam / np.sqrt(mem + 1e-8) # adagrad update

  p += seq_length # move data pointer
  n += 1 # iteration counter 

File: min-csv-rnn/csv-rnn.lua
-- Minimal implementation of vanilla RNN trained on csv data by Luca Naterop

function readCSV(file)
	-- Reads a csv file with comma delimiter and returns resulting tensor
	local csvFile = io.open(file, 'r')
	local firstLine = csvFile:read():split(',')
	local m = table.getn(firstLine)
	local N = csvFile:seek("end")/(2*m)
	csvFile:seek("set")        -- restore position

	local X = torch.zeros(N,m)
	local i = 0
	for line in csvFile:lines('*l') do
		i = i + 1
		local l = line:split(',')
		for key,val in ipairs(l) do
			X[i][key] = val
		end
	end
	return X
end

X = readCSV('Symbol.csv') 	-- training data
N = X:size(1) 				-- amount of frames
m = X:size(2)				-- amount of samples

print('Data has ' .. N .. ' frames and ' .. m .. ' samples')

-- hyperparameters
hidden_size 	= 130
seq_length 		= 25
learning_rate 	= 5e-2

-- model parameters
Wxh = 	torch.rand(hidden_size,m)*0.01
Whh = 	torch.rand(hidden_size,hidden_size)*0.01
Why = 	torch.rand(m,hidden_size)*0.01
bh = 	torch.zeros(hidden_size)
by = 	torch.zeros(m)

function lossFunction(inputs, targets, hprev)
	-- returns loss and gradients

	-- wichtig: hs[t] entspricht dem hs[t-1] im python code
	local xs 	= torch.zeros(seq_length, m)
	local hs 	= torch.zeros(seq_length+1, hidden_size)
	local ys 	= torch.zeros(seq_length, m)
	local diff 	= torch.zeros(seq_length, m)

	hs[1] = hprev:clone()
	local loss = 0
	-- forward propagation
	for t=1,seq_length do
		xs[t] 	= inputs[t]									-- fetch inputs
		hs[t+1] = torch.tanh(Wxh*xs[t] + Whh*hs[t] + bh) 	-- hidden layer activations
		ys[t] 	= Why*hs[t+1] + by							-- predictions
		diff[t] = targets[t] - ys[t]						-- deltas
		loss 	= loss + diff[t]:dot(diff[t])				-- squared error
	end

	--backward propagation
	local dWxh = 	torch.zeros(hidden_size,m)*0.01
	local dWhh = 	torch.zeros(hidden_size,hidden_size)*0.01
	local dWhy = 	torch.zeros(m,hidden_size)*0.01
	local dbh = 	torch.zeros(hidden_size)
	local dby = 	torch.zeros(m)
	local dhnext = 	torch.zeros(hidden_size) 

	for t = seq_length,1,-1 do
		dy 		= ys[t] - targets[t]
		dWhy 	= dWhy + torch.ger(dy,hs[t+1])
		dby 	= dby + dy
		dh 		= Why:t() * dy + dhnext 	-- backprop into h
		dhraw	= torch.cmul(dh, torch.pow(hs[t+1],2)*(-1)+1) -- backprop through tanh nonlinearity
		dbh		= dbh + dhraw
		dWxh 	= dWxh + torch.ger(dhraw,xs[t])
		dWhh 	= dWhh + torch.ger(dhraw,hs[t])
		dhnext 	= Whh:t()*dhraw
	end

	-- clip to mitigate exploding gradients
	dWxh:apply(clip)
	dWhh:apply(clip)
	dWhy:apply(clip)
	dbh:apply(clip)
	dby:apply(clip)

	return loss, dWxh, dWhh, dWhy, dbh, dby, hs[seq_length+1]
end

function sample(seed, h, amount)
	-- samples N vectors based on seed and hprev
	local samples = torch.zeros(amount+1,m)
	samples[1] = seed
	x = seed
	for i=2,amount+1 do
		h = torch.tanh(Wxh*x + Whh*h + bh)
		local y = Why*h + by
		samples[i] = y
		x = y
	end
	return samples
end

function clip(element)
	-- clips element to stay between [-5,5]
	if element > 5 then 
		return 5
	elseif element < -5 then 
		return -5
	else 
		return element
	end
end

function adagrad(Parameter, dParameter, Memory)
	-- performs adagrad on these tensors. Changes Param and Memory
	Memory = Memory + torch.cmul(dParameter,dParameter)
	Parameter = Parameter - torch.cdiv(dParameter,torch.sqrt(Memory + 1e-8))*learning_rate
	return Parameter, Memory
end

n = 0 -- iteration count
p = 1 -- data pointer
cycles = 0

-- memory variables for adagrad
mWxh = 	torch.zeros(hidden_size,m)
mWhh = 	torch.zeros(hidden_size,hidden_size)
mWhy = 	torch.zeros(m,hidden_size)
mbh = 	torch.zeros(hidden_size)
mby = 	torch.zeros(m)

smooth_loss = -math.log(1.0/m)*seq_length
hprev = torch.zeros(hidden_size) 
loss = 100

while true do
	-- check if we are at the end of training data
	if p + seq_length + 1 >= N or n == 0 then
		p = 1
		hprev = torch.zeros(hidden_size)
		cycles = cycles + 1
	end

	inputs = X:sub(p, p+seq_length-1)
	targets = X:sub(p+1, p+seq_length)

	if n % 1 == 0 then
		print('iter: '.. n ..', loss: '.. smooth_loss ..', pointer '.. p ..', cycles '..cycles)
	end

	loss, dWxh, dWhh, dWhy, dbh, dby, hprev = lossFunction(inputs, targets, hprev)
	smooth_loss = smooth_loss * 0.999 + loss * 0.001

  	Wxh, mWxh = adagrad(Wxh, dWxh, mWxh)
  	Whh, mWhh = adagrad(Whh, dWhh, mWhh)
  	Why, mWhy = adagrad(Why, dWhy, mWhy)
  	bh, mbh = adagrad(bh, dbh, mbh)
  	by, mby = adagrad(by, dby, mby)

	p = p + seq_length
	n = n + 1
	if cycles == 300 then
		sampleInput = X[p]
		break
	end
end

function writeCSV(M)
	local out = assert(io.open("./samples.csv", "w")) -- open a file for serialization
	splitter = ","
	for i=1,M:size(1) do
	    for j=1,M:size(2) do
	        out:write(M[i][j])
	        if j == M:size(2) then
	            out:write("\n")
	        else
	            out:write(splitter)
	        end
	    end
	end
	out:close()
end

local samples = sample(sampleInput, hprev, 400)

writeCSV(samples)
